### PR TITLE
Extra help for bad Java selections

### LINKED ISSUE #7386: Resolve inherited Java static method
Inherit static method from Java class is not recognized by Scala class
Here there is an small and simple example:

```scala
MyClass.java
public class MyClass extends MyClassBase {}

abstract class MyClassBase {
  public static void myStaticBaseMethod() {}
}

MyScalaClass.scala
class MyScalaClass {
  MyClassBase.myStaticBaseMethod() //compiles
  MyClass.myStaticBaseMethod() //no compiles. Wrong!!!
}
```

### PR DESCRIPTION
**Pull request by som-snytt**
In `J.m`, tell me if it would work in Java and how to say it in Scala.

Fixes scala/bug#7386

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala [hunk 1]
```diff
@@ -405,24 +403,22 @@ trait ContextErrors {
...
-              if (ObjectClass.info.member(name).exists) notAnyRefMessage(target)
-              else ""
-            )
-            companion + altStr + notAnyRef + semicolon
+            val semicolon = orEmpty(linePrecedes(qual, sel))(s"possible cause: maybe a semicolon is missing before `$nameString`?")
+            val notAnyRef = orEmpty(ObjectClass.info.member(name).exists)(notAnyRefMessage(target))
+            val javaRules = orEmpty(owner.isJavaDefined) {
+              def more = orEmpty(settings.explaintypes)(sm""" Static Java members belong to companion objects in Scala;
+                |they are not inherited, even by subclasses defined in Java.""")
+              val (jtype, jmember) = cx.javaFindMember(target, name, _ => true)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala]: Double check if this is prone to false positive, if it should check for static only, etc. Counterexamples in tests?

File: src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala [hunk 2]
```diff
@@ -283,9 +283,7 @@ trait TypeDiagnostics {
 
   // For found/required errors where AnyRef would have sufficed:
   // explain in greater detail.
-  def explainAnyVsAnyRef(found: Type, req: Type): String = {
-    if (AnyRefTpe <:< req) notAnyRefMessage(found) else ""
-  }
+  def explainAnyVsAnyRef(found: Type, req: Type): String = if (AnyRefTpe <:< req) s"\n${notAnyRefMessage(found)}" else ""
```
Comment (hrhino) [inline src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala]: Not `orEmpty`?

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala]: Something like that. Only prepend nl if nonEmpty.

File: src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala [hunk 3]
```diff
@@ -405,24 +403,22 @@ trait ContextErrors {
...
-            )
-            val notAnyRef = (
-              if (ObjectClass.info.member(name).exists) notAnyRefMessage(target)
-              else ""
-            )
-            companion + altStr + notAnyRef + semicolon
+            val semicolon = orEmpty(linePrecedes(qual, sel))(s"possible cause: maybe a semicolon is missing before `$nameString`?")
+            val notAnyRef = orEmpty(ObjectClass.info.member(name).exists)(notAnyRefMessage(target))
+            val javaRules = orEmpty(owner.isJavaDefined && owner.isClass && !owner.isPackage) {
+              def more = orEmpty(settings.explaintypes)(sm""" Static Java members belong to companion objects in Scala;
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala]: I wouldn't hide this explanation behind `-explaintypes`, it's too helpful and non-confusing.

**Comment by som-snytt** [comment]
Did not attempt to re-architect "How do I append words to the current user message."

**Review by lrytz — APPROVED** [review]
Thanks!

**Comment by som-snytt** [comment]
Thanks @retronym for his commitment to Java interop!

**Pull request merged**